src/aging_session/memory.clj

```diff
@@ -43,6 +64,67 @@
     (assoc session-map key entry)
     (dissoc session-map key)))
 
+(def ^:private internals-atom (atom {:enc-key nil
+                                     :mac-key nil
+                                     :counter 0}))
+
+(defn get-crypto-keys
+  []
+  (let [{enc-key :enc-key
+         mac-key :mac-key} @internals-atom]
+    (if (and enc-key mac-key)                               ; if keys are present - use them
+      [enc-key mac-key]
+      (do                                                   ; otherwise generate new keys and attempt to set them
+        (swap! internals-atom update-in [:enc-key] #(if-not % (nonce/random-bytes 32) %))
+        (swap! internals-atom update-in [:mac-key] #(if-not % (nonce/random-bytes 128) %))
+        (get-crypto-keys)))))                               ; retrieve the keys again, as something else may have set them in the meantime
+
+(defn decrypt
+  [iv key data]
+  (try
+    (nippy/thaw (crypto/decrypt-cbc (crypto/block-cipher :aes :cbc) data key iv))
+    (catch Exception e
+      (log/error "Error decrypting cookie: " e)
+      nil)))
+
+(defn encrypt
+  [key data]
+  (let [iv (nonce/random-bytes 16)]
+    [iv (crypto/encrypt-cbc (crypto/block-cipher :aes :cbc) (nippy/freeze data) key iv)]))
+
+(defn verify-and-decrypt-cookie
+  [url-cookie]
```

Comment:
will do

Reply:
Regarding your question on Slack about the example I gave for `enc-cookie` : what I meant is not exactly more than one purpose. Yeah, it always represents the `enc-cookie`, so it's does not have more than one purpose. 
What I meant is that binding a value to the same name again and again is seen as not idiomatic in Clojure and functional programming as a whole.
Clojure lets you do it - I didn't even know it could be done because more purely functional languages don't! - but this is considered bad practice. Clojure struggles to maintain mutability to a certain part of it and even has structures - like atoms - that are made specifically for cases where you want to mutate stuff. 
So doing:
```
(let [x 1
        x (+ 1 x)]
        (println x))
``` 
Is considered bad practice and is not even allowed in more purely functional languages. And Clojure let's you do it - God knows why! - but the idea is not to from a functional point of view.
What you should be doing is:

```
(let [x1 1
        x2 (+ 1 x1)]
        (println y))
``` 
Avoiding mutability in functional programming when it's permitted - like in Clojure - is also done because it helps you write a code that is easier to debug. In this second example of code, you could inspect the intermediate steps of your calculation more easily than in the first one example.

There's a similar discussion about it here:
https://softwareengineering.stackexchange.com/questions/306889/if-you-can-use-def-to-redefine-variables-how-is-that-considered-immutable

You can also look for "Imutability in functional programming" to see more discussion about it.